chia/data_layer/data_layer_client.py

```diff
@@ -0,0 +1,117 @@
+            print(f"Got root hash: {node}")
+            t1 = time.time()
+            internal_nodes = 0
+            terminal_nodes = 0
+            stack: List[str] = []
+            add_to_db_cache: Dict[str, Any] = {}
+            while node is not None:
+                url = f"http://0.0.0.0:8080/get_tree_nodes?tree_id={tree_id}&node_hash={node}&root_hash={root_hash}"
+                async with session.get(url) as r:
+                    json = await r.json()
+                root_changed = json["root_changed"]
+                if root_changed:
+                    print("Data changed since the download started. Aborting.")
+                    return
+                answer = json["answer"]
+                for row in answer:
+                    if row["is_terminal"]:
+                        key = row["key"]
+                        value = row["value"]
+                        hash = Program.to((hexstr_to_bytes(key), hexstr_to_bytes(value))).get_tree_hash()
+                        if hash.hex() == node:
+                            if verbose:
+                                print(f"Received terminal node {key} {value}.")
+                            await self.data_store._insert_node(node, NodeType.TERMINAL, None, None, key, value)
+                            if verbose:
+                                print(f"Added terminal node {hash} to DB.")
+                            terminal_nodes += 1
+                            right_hash = hash.hex()
+                            while right_hash in add_to_db_cache:
+                                node, left_hash = add_to_db_cache[right_hash]
+                                del add_to_db_cache[right_hash]
+                                await self.data_store._insert_node(
+                                    node, NodeType.INTERNAL, left_hash, right_hash, None, None
+                                )
+                                internal_nodes += 1
+                                if verbose:
+                                    print(f"Added internal node {node} to DB.")
+                                right_hash = node
+                        else:
+                            raise RuntimeError(
+                                f"Did not received expected node. Expected: {node} Received: {hash.hex()}"
+                            )
+                        if len(stack) > 0:
+                            node = stack.pop()
+                        else:
+                            node = None
+                    else:
+                        left_hash = row["left"]
+                        right_hash = row["right"]
+                        left_hash_bytes = hexstr_to_bytes(left_hash)
+                        right_hash_bytes = hexstr_to_bytes(right_hash)
+                        hash = Program.to((left_hash_bytes, right_hash_bytes)).get_tree_hash(
+                            left_hash_bytes, right_hash_bytes
+                        )
```

Comment:
this doesn't look right. it looks like you're "flattening" the tree and compute the tree hash of a node of hashes, not a node of subtrees. isn't there a function to compute the tree hash of the data store that you can use here instead of sticking this algorithm in here directly? Either way, I think it should be factored out into a function so you can unit test it.

Reply:
I have used the method from here:
https://github.com/Chia-Network/chia-blockchain/blob/atari/chia/data_layer/data_store.py#L143
but perhaps factor it out and call the same function in both places?